File: ios/Message/TXIMMessageBuilder.m
//
//  TXIMMessageBuilder.m
//  react-native-tencent-im
//
//  Created by Choi Wing Chiu on 22/8/2020.
//

#import "TXIMMessageBuilder.h"

#define CURRENT_TIMESTAMP [[NSDate date] timeIntervalSince1970] * 1000

@implementation TXIMMessageBuilder



+ (TXIMMessageInfo *)buildTextMessage:(NSString *)content {
    TXIMMessageInfo *info = [[TXIMMessageInfo alloc] initWithType:TXIMMessageTypeText];
    V2TIMMessage *msg = [[V2TIMManager sharedInstance] createTextMessage:content];
    info.msg = msg;
    info.msgTime = CURRENT_TIMESTAMP;
    info.extra = content;
    info.isSelf = YES;
    info.sender = [[V2TIMManager sharedInstance] getLoginUser];
    return info;
}

+ (TXIMMessageInfo *)buildCustomFaceMessage:(NSString *)groupId faceName:(NSString *)faceName {
    TXIMMessageInfo *info = [[TXIMMessageInfo alloc] initWithType:TXIMMessageTypeFace];
    V2TIMMessage *msg = [[V2TIMManager sharedInstance] createFaceMessage:[groupId intValue] data:[faceName dataUsingEncoding:NSUTF8StringEncoding]];
    info.msg = msg;
    info.msgTime = CURRENT_TIMESTAMP;
    info.extra = @"[自定义表情]";
    info.isSelf = YES;
    info.sender = [[V2TIMManager sharedInstance] getLoginUser];
    return info;
}

+ (TXIMMessageInfo *)buildImageMessage:(NSString *)uri compressed:(Boolean *)compressed appPohto:(Boolean *)appPohto {
    TXIMMessageInfo *info = [[TXIMMessageInfo alloc] initWithType:TXIMMessageTypeImage];
    V2TIMMessage *msg = [[V2TIMManager sharedInstance] createImageMessage:uri];
    info.msg = msg;
    info.msgTime = CURRENT_TIMESTAMP;
    info.extra = @"[图片]";
    info.isSelf = YES;
    info.sender = [[V2TIMManager sharedInstance] getLoginUser];
    return info;
}


+ (TXIMMessageInfo *)buildMessage:(TXIMMessageType)type content:(NSString *)content option:(NSDictionary *)option {
    switch (type) {
        case TXIMMessageTypeText:
            return [self buildTextMessage: content];
        default:
            return nil;
    }
}

+ (TXIMMessageInfo *)buildMessageWithTIMMessage:(V2TIMMessage *)msg {
    if (!msg || msg.status == V2TIM_MSG_STATUS_HAS_DELETED) {
        return nil;
    }
    TXIMMessageInfo *info = [TXIMMessageInfo new];
    [info setMsgType:[self getMessagetype:msg.elemType]];
    [info setMsg:msg];
    [info setMsgId:[msg msgID]];
    [info setMsgTime:[[msg timestamp] timeIntervalSince1970] * 1000];
    [info setIsSelf:[msg isSelf]];

    if (info.msgType == TXIMMessageTypeText) {
        info.extra = msg.textElem.text;
    }
    
    if (info.isSelf) {
        info.sender = [msg sender];
        info.receiver = [msg userID];
        info.isRead = [msg isPeerRead];
    } else {
        info.sender = [msg sender];
        info.senderAvatar = [msg faceURL];
        info.senderNickName = [msg nickName];
    }
    return info;
}

+ (TXIMMessageType) getMessagetype:(V2TIMElemType)type {
    switch (type) {
        case V2TIM_ELEM_TYPE_NONE:
            return TXIMMessageTypeunknow;
        case V2TIM_ELEM_TYPE_TEXT:
            return TXIMMessageTypeText;
        case V2TIM_ELEM_TYPE_CUSTOM:
            return TXIMMessageTypeCustom;
        case V2TIM_ELEM_TYPE_IMAGE:
            return TXIMMessageTypeImage;
        case V2TIM_ELEM_TYPE_SOUND:
            return TXIMMessageTypeAudio;
        case V2TIM_ELEM_TYPE_VIDEO:
            return TXIMMessageTypeVideo;
        case V2TIM_ELEM_TYPE_FILE:
            return TXIMMessageTypeFile;
        case V2TIM_ELEM_TYPE_LOCATION:
            return TXIMMessageTypeLocation;
        case V2TIM_ELEM_TYPE_FACE:
            return TXIMMessageTypeFace;
        case V2TIM_ELEM_TYPE_GROUP_TIPS:
            return TXIMMessageTypeGroupTips;
        default:
            return TXIMMessageTypeunknow;
    }
}

+ (NSArray<TXIMMessageInfo *>*) normalizeConversationList:(NSArray<V2TIMConversation*> *)timMessages {
    NSMutableArray *messageInfos = [[NSMutableArray alloc] init];
    for (int i = 0 ; i < timMessages.count ; ++ i) {
        NSMutableDictionary *objmap = [[NSMutableDictionary alloc] init];
        V2TIMConversation *conv = timMessages[i];
        int unreadNum = [conv unreadCount];
        V2TIMConversationType type = [conv type];
        V2TIMMessage *msg = [conv lastMessage];
        TXIMMessageInfo *info = [self buildMessageWithTIMMessage:msg];
        NSString *peer = [conv userID] != nil ? [conv userID] : [conv groupID];
        [objmap setValue:[NSString stringWithFormat:@"%d", unreadNum] forKey:@"unread"];
        if (!!msg) {
            [objmap setValue:[info toDict] forKey:@"message"];
        }
        [objmap setValue:peer forKey:@"peer"];
        if (type == V2TIM_C2C) {
            [objmap setValue:@"1" forKey:@"type"];
        } else {
            [objmap setValue:@"2" forKey:@"type"];
        }
        [objmap setValue:[conv showName] forKey:@"name"];
        [messageInfos addObject:objmap];
    }
    return messageInfos;
}

+ (NSArray<TXIMMessageInfo *>*) TIMMessages2MessageInfos:(NSArray<V2TIMConversation*> *) timMessages isGroup:(Boolean *)isGroup {
    NSMutableArray *messageInfos = [[NSMutableArray alloc] init];
    
    return messageInfos;
}

//+ (TXIMMessageInfo *) TIMMessage2MessageInfo:(V2TIMMessage *)msg isGroup:(Boolean *)isGroup {
//    
//}

@end

File: android/src/main/java/com/chiu/message/TXIMMessageBuilder.java
package com.chiu.tencentim.message;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;
import com.tencent.imsdk.v2.V2TIMConversation;
import com.tencent.imsdk.v2.V2TIMElem;
import com.tencent.imsdk.v2.V2TIMImageElem;
import com.tencent.imsdk.v2.V2TIMManager;
import com.tencent.imsdk.v2.V2TIMMessage;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;
import java.util.ListIterator;

public class TXIMMessageBuilder {
    public static TXIMMessageInfo buildTextMessage(String message) {
        TXIMMessageInfo info = new TXIMMessageInfo();
        V2TIMMessage msg = V2TIMManager.getMessageManager().createTextMessage(message);
        info.setExtra(message);
        info.setMsgTime(System.currentTimeMillis());
        info.setSelf(true);
        info.setMsg(msg);
        info.setSender(V2TIMManager.getInstance().getLoginUser());
        info.setMsgType(TXIMMessageInfo.MSG_TYPE_TEXT);
        return info;
    }

    public static TXIMMessageInfo buildMessage(int type, String content) {
        switch (type) {
            case TXIMMessageInfo.MSG_TYPE_TEXT:
                return buildTextMessage(content);
            default:
                return null;
        }
    }

    public static TXIMMessageInfo buildMessageWithTIMMessage(V2TIMMessage msg) {
        if (msg == null || msg.getStatus() == V2TIMMessage.V2TIM_MSG_STATUS_HAS_DELETED) {
            return null;
        }
        TXIMMessageInfo info = new TXIMMessageInfo();
        info.setMsg(msg);
        info.setMsgId(msg.getMsgID());
        info.setMsgTime(msg.getTimestamp() * 1000);
        info.setSelf(msg.isSelf());
        if (msg.getElemType() == V2TIMMessage.V2TIM_ELEM_TYPE_TEXT) {
            info.setExtra(msg.getTextElem().getText());
        }
        if (info.isSelf()) {
            info.setSender(msg.getSender());
            info.setReceiver(msg.getUserID());
            info.setRead(msg.isPeerRead());
        } else {
            info.setSender(msg.getSender());
            info.setSenderAvatar(msg.getFaceUrl());
            info.setSenderNickName(msg.getNickName());
        }
        return info;
    }

    public static int getMessageType(int type) {
        switch (type) {
            default:
            case V2TIMMessage.V2TIM_ELEM_TYPE_NONE:
                return TXIMMessageInfo.MSG_TYPE_UNKNOW;
            case V2TIMMessage.V2TIM_ELEM_TYPE_TEXT:
                return TXIMMessageInfo.MSG_TYPE_TEXT;
            case V2TIMMessage.V2TIM_ELEM_TYPE_CUSTOM:
                return TXIMMessageInfo.MSG_TYPE_CUSTOM;
            case V2TIMMessage.V2TIM_ELEM_TYPE_IMAGE:
                return TXIMMessageInfo.MSG_TYPE_IMAGE;
            case V2TIMMessage.V2TIM_ELEM_TYPE_SOUND:
                return TXIMMessageInfo.MSG_TYPE_AUDIO;
            case V2TIMMessage.V2TIM_ELEM_TYPE_VIDEO:
                return TXIMMessageInfo.MSG_TYPE_VIDEO;
            case V2TIMMessage.V2TIM_ELEM_TYPE_FILE:
                return TXIMMessageInfo.MSG_TYPE_FILE;
            case V2TIMMessage.V2TIM_ELEM_TYPE_LOCATION:
                return TXIMMessageInfo.MSG_TYPE_LOCATION;
            case V2TIMMessage.V2TIM_ELEM_TYPE_FACE:
                return TXIMMessageInfo.MSG_TYPE_CUSTOM_FACE;
            case V2TIMMessage.V2TIM_ELEM_TYPE_GROUP_TIPS:
                return TXIMMessageInfo.MSG_TYPE_GROUP_TIPS;
        }
    }

    public static List<TXIMMessageInfo> normalizeMessageHistory(List<V2TIMMessage> messages) {
        ArrayList<TXIMMessageInfo> list = new ArrayList<TXIMMessageInfo>();
        for (int i = 0 ; i < messages.size() ; ++ i) {
            V2TIMMessage msg = messages.get(i);
            TXIMMessageInfo info = TXIMMessageBuilder.buildMessageWithTIMMessage(msg);
            list.add(info);
        }
        return list;
    }

    public static WritableArray normalizeConversationList(List<V2TIMConversation> timMessages) {
        WritableArray list = Arguments.createArray();
        for (int i = 0 ; i < timMessages.size() ; ++ i) {
            WritableMap obj = Arguments.createMap();
            V2TIMConversation conv = timMessages.get(i);
            int unreadNum = conv.getUnreadCount();
            V2TIMMessage msg = conv.getLastMessage();
            TXIMMessageInfo info = TXIMMessageBuilder.buildMessageWithTIMMessage(msg);
            String peer = conv.getGroupID() == null ? conv.getUserID() : conv.getGroupID();
            obj.putInt("unread", unreadNum);
            if (msg != null) {
                obj.putMap("message", info.toDict());
            }
            obj.putString("peer", peer);
            if (conv.getType() == 1) {
                obj.putString("type", "1");
            } else {
                obj.putString("type", "2");
            }
            obj.putString("name", conv.getShowName());
            obj.putString("faceUrl", conv.getFaceUrl());
            list.pushMap(obj);
        }
        return list;
    }
}
